### PR TITLE
New CLI subcommand to create clang-compatible compilation database (`compile_commands.json`)

### LINKED ISSUE #4855: YouCompleteMe support for auto completion
This is not an issue, but utilizing some other program to make editing easier; and if it can be enabled easily from the upstream.

## Feature Request Type

- [ ] Core Functionality
- [ ] Add-on hardware support (e.g. audio, RGB, OLED screen, etc.)
- [ ] Alteration (enhancement/optimization) of existing Feature(s)
- [x] New behavior

## Description 

I am still a beginner when it comes to C. I use vim to edit files, with [YouCompleteMe](https://github.com/Valloric/YouCompleteMe) plugin for autocompletion.

YouCompleteMe requires a [config file, or a compilation database](https://github.com/Valloric/YouCompleteMe#c-family-semantic-completion) for completion, but I am unsure how to get that working by myself. 

What I don't understand is; can this be handled by compilation database (modification in make?) or do I need to specify myself on a config file.

### PR DESCRIPTION
**Pull request by qubidt**
## Description

Updating @xton's PR #10264, which adds a CLI command to generate a [Clang Compilation Database file](https://clang.llvm.org/docs/JSONCompilationDatabase.html), for use with language servers, linters, and other tooling. No major changes other than addressing merge conflicts.

I changed the subcommand name, as was requested in the review for the previous PR, but to `generate-compilation-database` instead of `generate-compile-commands` as @skullydazed suggested. This seems more aligned with how the file is referred to in compatible tools' documentation. But I can change this if needed.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes #4855

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tzarc** [comment]
Any time I try to execute it gives this error, regardless of the parameters:
```☒ Could not determine keyboard!
☒ You must supply both `--keyboard` and `--keymap`, or be in a directory for a keyboard or keymap.```

**Comment by qubidt** [comment]
> Any time I try to execute it gives this error, regardless of the parameters

Hmm, I have those settings saved to my config so I didn't notice that. I'll figure it out

> You'll also need to remove the debugging statements when printing out loading of modules

agh I didn't realize I committed those print statements. that's embarrassing 😅

**Comment by qubidt** [comment]
Fixed the issue. I missed the config parameters when I renamed the subcommand, so they weren't being parsed correctly.

Also added error handling for the case where no commands were generated (e.g. you specified a non-existent keymap) and some minor improvements (instead of checking checking exit status or calling `which ...` ourselves, delegate it to `Popen`/`shutil`, respectively)

File: lib/python/qmk/cli/generate/compilation_database.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        env = os.environ.copy()
+        env.pop("MAKEFLAGS", None)
+
+        # re-use same executable as the main make invocation (might be gmake)
+        clean_command = [command[0], 'clean']
+        cli.log.info('Making clean with {fg_cyan}%s', ' '.join(clean_command))
+        subprocess.run(clean_command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, env=env, check=True)
+
+        cli.log.info('Gathering build instructions from {fg_cyan}%s', ' '.join(command))
+        with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, env=env) as proc:
```
Comment (skullydazed) [inline lib/python/qmk/cli/generate/compilation_database.py:107]: I don't have time to do a full review right now, but please use `cli.run` here as it handles running commands on windows correctly. Using subprocess directly in a way that works cross-platform is error-prone.

Comment (qubidt) [reply lib/python/qmk/cli/generate/compilation_database.py:107]: Good call, I wasn't aware of milc's `cli.run`. That'll be straightforward to fix.

As for cross-platform compatibility, I wasn't thinking of that at all. I'll have to spin up a windows environment to test it, but I don't think the way @xton searched for the system library path is very portable. Not sure how I"ll tackle that yet but any suggestions are welcome.

**Comment by qubidt** [comment]
Now using `cli.run()` instead of `subprocess.Popen()`. It now waits for the command to complete before parsing the output from `make` but there's relatively little output so the impact to overall run time is negligible.

We will still likely need to fix the `system_libs()` function to give reasonable output for windows environments. I literally never use windows so I'll have to do some research to figure out what that will look like.

**Comment by AWare** [comment]
Just a quick note of thanks, this PR has helped me make vscode a lot happier about qmk 👏

**Review by tzarc — APPROVED** [review]

**Comment by Diaoul** [comment]
Trying to use this I still have a few issues in vscode in keymaps after generating the database:
```
Unknown type name 'uint16_t' (fix available) clang(unknown_typename)
```
and 
```
Expected ';' after top level declarator (fix available) clang(invalid_token_after_toplevel_declarator)
```

Do I need specific clangd options or anything I am missing? Compiling an flashing work fine though :thinking:

**Comment by user78552** [comment]
@Diaoul Hi, did you manage to solve this problem?

**Comment by Diaoul** [comment]
No, I gave up.

**Pull request merged**